### PR TITLE
[catnap] Enhancement: Move to socket2

### PR DESCRIPTION
**Pull request by iyzhang**
This code is a merge of the Windows and Linux code paths for Catnap into a single socket2 API. It is a first step to supporting epoll and completion ports. There are 3 commits: 
1. Just commenting out some overwhelming debug messages.
2. The bulk of the changes adding a CatnapTransport
3. An extra bug fix. It seems like if we start the next test too soon after the previous one exits, the old server address is still bound for a little while. We should probably set to this to some reasonable number but not really sure what.
4. Remove level of unnecessary abstraction for sockets. Now that we are using Socket2, we can keep all of the per-queue code in either the CatnapQueue or the transport.

### CODE REVIEW

File: src/rust/demikernel/libos/network.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ impl NetworkLibOS {
             #[cfg(feature = "catpowder-libos")]
             NetworkLibOS::Catpowder(libos) => libos.socket(domain, socket_type, protocol),
             #[cfg(all(feature = "catnap-libos"))]
-            NetworkLibOS::Catnap(libos) => libos.socket(domain, socket_type, protocol),
+            NetworkLibOS::Catnap(libos) => libos.socket(domain.into(), socket_type.into(), protocol.into()),
```
Comment (ppenna) [inline src/rust/demikernel/libos/network.rs:69]: Once you move all LibOSes to this, I would sugges us to move this conversion logic up to the PDPIX layer. What do you think?

**Comment by ppenna** [comment]
> An extra bug fix. It seems like if we start the next test too soon after the previous one exits, the old server address is still bound for a little while. We should probably set to this to some reasonable number but not really sure what.

@iyzhang this is due to the "TIME_WAIT" state. There are some per-system solutions that might be explored, such as using the `SO_REUSEADDR`.

File: tests/rust/tcp-test/bind/mod.rs [hunk 2]
```diff
@@ -245,7 +245,7 @@ fn bind_to_non_local_address(libos: &mut LibOS) -> Result<()> {
 
     // Fail to bind socket.
     match libos.bind(sockqd, addr) {
-        Err(e) if e.errno == libc::EADDRNOTAVAIL => (),
+        Err(e) if e.errno == libc::EADDRNOTAVAIL || e.errno == libc::EBADF => (),
```
Comment (ppenna) [inline tests/rust/tcp-test/bind/mod.rs:248]: Where is this bug comming from?

File: src/rust/catnap/queue.rs [hunk 3]
```diff
@@ -293,6 +369,32 @@ impl SharedCatnapQueue {
...
+    /// Generic function for spawning a control-path coroutine on [self].
+    fn do_generic_sync_control_path_call<F>(
+        &mut self,
+        coroutine_constructor: F,
+        yielder: Yielder,
+    ) -> Result<TaskHandle, Fail>
+    where
+        F: FnOnce(Yielder) -> Result<TaskHandle, Fail>,
+    {
+        // Spawn coroutine.
```
Comment (anandbonde) [inline src/rust/catnap/queue.rs:381]: With the new name `coroutine_constructor`, this comment probably needs to be revisited.

**Review by anandbonde — APPROVED** [review]
Looks good.

File: tests/rust/tcp-test/bind/mod.rs (hunk 2 above)
Comment (iyzhang) [reply tests/rust/tcp-test/bind/mod.rs:248]: socket2 returns EBADF instead from this call.

**Comment by iyzhang** [comment]
> > An extra bug fix. It seems like if we start the next test too soon after the previous one exits, the old server address is still bound for a little while. We should probably set to this to some reasonable number but not really sure what.
> 
> @iyzhang this is due to the "TIME_WAIT" state. There are some per-system solutions that might be explored, such as using the `SO_REUSEADDR`.

Ah great! Let me know what the right solution is. I can certainly set SO_REUSEADDR on all listening sockets.

File: src/rust/demikernel/libos/network.rs (hunk 1 above)
Comment (iyzhang) [reply src/rust/demikernel/libos/network.rs:69]: Yes, definitely! I would love to get rid of some of this code at the libos/network.rs level

File: tests/rust/tcp-test/bind/mod.rs (hunk 2 above)
Comment (ppenna) [reply tests/rust/tcp-test/bind/mod.rs:248]: Sure, but then this does not comply with expected behavior for sockets. Shouldn't we thus interpose the returned error code?

**Comment by ppenna** [comment]
> > > An extra bug fix. It seems like if we start the next test too soon after the previous one exits, the old server address is still bound for a little while. We should probably set to this to some reasonable number but not really sure what.
> > 
> > 
> > @iyzhang this is due to the "TIME_WAIT" state. There are some per-system solutions that might be explored, such as using the `SO_REUSEADDR`.
> 
> Ah great! Let me know what the right solution is. I can certainly set SO_REUSEADDR on all listening sockets.

Setting `SO_REUSEADDR` in all listening sockets works for Linux, not sure how that behaves in Windows. @kyleholohan any ideas on this?

File: tests/rust/tcp-test/bind/mod.rs (hunk 2 above)
Comment (iyzhang) [reply tests/rust/tcp-test/bind/mod.rs:248]: Oh nooo

File: src/rust/catnap/queue.rs (hunk 3 above)
Comment (iyzhang) [reply src/rust/catnap/queue.rs]: Done

File: tests/rust/tcp-test/bind/mod.rs (hunk 2 above)
Comment (iyzhang) [reply tests/rust/tcp-test/bind/mod.rs:248]: I'm not sure how to distinguish it from the socket being closed if I just translate all EBADF to EADDRNOTAVAIL

Comment (iyzhang) [reply tests/rust/tcp-test/bind/mod.rs:248]: Let me revisit this if/when we move to TcpStream

**Pull request merged**